IPython/kernel/zmq/kernelbase.py

```diff
@@ -347,41 +347,43 @@ def execute_request(self, stream, ident, parent):
             self.log.error("Got bad msg: ")
             self.log.error("%s", parent)
             return
-        
-        stop_on_error = content.get('stop_on_error', True)
 
         md = self._make_metadata(parent['metadata'])
-        
+
         # Re-broadcast our input for the benefit of listening clients, and
         # start computing output
         if not silent:
             self.execution_count += 1
             self._publish_execute_input(code, parent, self.execution_count)
-        
+
         reply_content = self.do_execute(code, silent, store_history,
                                         user_expressions, allow_stdin)
+        self.send_execute_reply(stream, ident, parent, md, reply_content)
+
+    def send_execute_reply(self, stream, ident, parent, md, reply_content):
+        """ Send a reply to execute_request """
 
         # Flush output before sending the reply.
-        sys.stdout.flush()
-        sys.stderr.flush()
-        # FIXME: on rare occasions, the flush doesn't seem to make it to the
-        # clients... This seems to mitigate the problem, but we definitely need
-        # to better understand what's going on.
-        if self._execute_sleep:
-            time.sleep(self._execute_sleep)
+        self._flush_execute_output()
 
-        # Send the reply.
+        # Cleanup the reply and prepare metadata
         reply_content = json_clean(reply_content)
-        
+
+        md = self._make_metadata(md)
         md['status'] = reply_content['status']
         if reply_content['status'] == 'error' and \
                         reply_content['ename'] == 'UnmetDependency':
                 md['dependencies_met'] = False
 
+        # Publish the reply
         reply_msg = self.session.send(stream, u'execute_reply',
                                       reply_content, parent, metadata=md,
                                       ident=ident)
-        
+        # Handle the reply message
+        content = parent[u'content']
+        stop_on_error = content.get('stop_on_error', True)
+        silent = content[u'silent']
```

Comment:
`stop_on_error` and `silent` are only used once below, it's probably not necessary to assign them to variables again.